pkg/login/social/azuread_oauth.go

```diff
@@ -160,6 +115,65 @@ func (s *SocialAzureAD) UserInfo(ctx context.Context, client *http.Client, token
 	}, nil
 }
 
+func (s *SocialAzureAD) validateClaims(ctx context.Context, client *http.Client, parsedToken *jwt.JSONWebToken) (*azureClaims, error) {
+	claims, err := s.validateIDTokenSignature(ctx, client, parsedToken)
+	if err != nil {
+		return nil, fmt.Errorf("error getting claims from id token: %w", err)
+	}
+
+	if claims.OAuthVersion == "1.0" {
+		return nil, &Error{"AzureAD OAuth: version 1.0 is not supported. Please ensure the auth_url and token_url are set to the v2.0 endpoints."}
+	}
+
+	s.log.Debug("Validating audience", "audience", claims.Audience, "client_id", s.ClientID)
+	if claims.Audience != s.ClientID {
+		return nil, &Error{"AzureAD OAuth: audience mismatch"}
+	}
+
+	s.log.Debug("Validating tenant", "tenant", claims.TenantID, "allowed_tenants", s.allowedOrganizations)
+	if !s.isAllowedTenant(claims.TenantID) {
+		return nil, &Error{"AzureAD OAuth: tenant mismatch"}
+	}
+	return claims, nil
+}
+
+func (s *SocialAzureAD) validateIDTokenSignature(ctx context.Context, client *http.Client, parsedToken *jwt.JSONWebToken) (*azureClaims, error) {
+	var claims azureClaims
+
+	jwksFuncs := []func(ctx context.Context, client *http.Client, authURL string) (*keySetJWKS, time.Duration, error){
+		s.retrieveJWKSFromCache, s.retrieveSpecificJWKS, s.retrieveGeneralJWKS,
+	}
+
+	keyID := parsedToken.Headers[0].KeyID
+
+	for _, jwksFunc := range jwksFuncs {
+		keyset, expiry, err := jwksFunc(ctx, client, s.Endpoint.AuthURL)
+		if err != nil {
+			return nil, fmt.Errorf("error retrieving jwks: %w", err)
+		}
+		var errClaims error
+		keys := keyset.Key(keyID)
+		for _, key := range keys {
+			s.log.Debug("AzureAD OAuth: trying to parse token with key", "kid", key.KeyID)
+			if errClaims = parsedToken.Claims(key, &claims); errClaims == nil {
+				if expiry != 0 {
+					s.log.Debug("AzureAD OAuth: caching key set", "kid", key.KeyID, "expiry", expiry)
+					if err := s.cacheJWKS(ctx, keyset, expiry); err != nil {
```

Comment:
1. It's either caches the default or the `?appid=` keyset, am I correct?
2. I would create a function that executes `retrieveGeneralJWKS` and `retrieveSpecificJWKS`, merge those and cache it with expiration=min(expGeneralJWKS, expSpecificJWKS), then the cache has all the keys in case those are needed for other verification and this function could be slimmer.

What do you think?